# Table of Contents

0.1 Importing Libraries

0.2 Importing Data

0.3 Exploring Original Dfs

0.4 OUTER Merging Dfs

0.5 Exploring Merged Df


### 0.1 Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 0.2 Importing Data

In [2]:
# Identify the file pathway to data files
path = r'C:\Users\CJ\Documents\_CJ-Stuff\Career Foundry\Data Immersion\Ach 4 - Python\2023-03 Instacart Basket Analysis'

In [3]:
# Import data from products_clean and orders_products_combined.csv

In [4]:
df_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'products_clean.pkl'))

In [5]:
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

### 0.3 Exploring Original Dfs

In [6]:
# Comparing shapes of the two dfs
df_orders_products_combined.shape

(32434489, 10)

In [7]:
df_prods.shape

(49668, 5)

Merge should be 32434489 rows and 15 columns.
(Columns = 10 + 5 + 1 merge field - 1 key overlap)

In [8]:
# Check the output of the two dfs
df_orders_products_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  
0                  1          0  
1                  2          0  
2                  3          0  
3                  4          0  
4                  5          0

In [9]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  price  
0             19    5.8  
1             13    9.3  
2              7    4.5  
3              1   10.5  
4             13    4.3

In [10]:
# Compare the datatypes
df_orders_products_combined.dtypes

order_id                         int64
user_id                          int64
order_number                     int16
order_day_of_week                 int8
order_hour_of_day                 int8
days_since_prior_order         float16
first_order_of_new_customer       bool
product_id                       int32
add_to_cart_order                int16
reordered                         int8
dtype: object

In [11]:
df_prods.dtypes

product_id         int32
product_name      object
aisle_id           int16
department_id      int16
price            float32
dtype: object

The product_id column will be the key for merging the two tables.

### 0.4 OUTER Merging Dfs

In [12]:
# An exploratory OUTER merge 
df_merged = df_orders_products_combined.merge(df_prods, on = ['product_id'], how = 'outer', indicator = True)

### 0.5 Exploring Merged Df

In [16]:
df_merged['_merge'].value_counts()

both          32403719
left_only        30770
right_only          11
Name: _merge, dtype: int64

Most rows found a match in both tables.

30,770 orders are for products where there is no product info.

11 products have never been ordered. 

This means we should do a LEFT merge the final merge because we want to be able explore all orders -- even if we aren't sure what product was purchased.
 

In [15]:
df_merged.shape

(32434500, 15)

A perfect inner merge would have 32,434,489 rows and 15 columns. 

Given the value_counts of the outer merge, we'd expect to see 32,434,500
